### PR TITLE
Mantin camera handler

### PR DESCRIPTION
**Pull request by uriahrokach**

### CODE REVIEW

File: src/com/spikes2212/Cameras/CameraController.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package com.spikes2212.Cameras;
+
+import java.util.Optional;
+
+import com.ni.vision.NIVision.Image;
+
+import edu.wpi.first.wpilibj.vision.USBCamera;
+
+public class CameraController {
+	public static int resolutionWidth = 100, resolutionHeight = 100;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CameraController.java]: Make final and give reasonable values, also add a constructor in which you can specify those

File: src/com/spikes2212/Cameras/CameraController.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	public void stop() {
+		try {
+			camera.ifPresent(USBCamera::stopCapture);
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+		on = false;
+	}
+
+	public void getImage(Image image) {
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CameraController.java]: add check to avoid reading from camera if it's turned of, instead of unnecessary try-catch

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.ni.vision.NIVision.Image;
+import com.ni.vision.NIVision.ImageType;
+
+import edu.wpi.first.wpilibj.CameraServer;
+
+public class CamerasHandler {
+	private ArrayList<CameraController> cameras;
+	private CameraController currentCamera = null;
+	private Image image;
+	public static int fps = 30;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: make final, also add constructor that allows you to specify the FPS

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	public void addCamera(String port) {
+		cameras.add(new CameraController(port, fps));
+	}
+
+	public synchronized void stop() {
+		cameras.forEach(c -> c.stop());
+		currentCamera = null;
+	}
+
+	public synchronized void start(int index) {
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: either allow you to start a camera based on its port, or change it so when you add a camera you can specify the index it goes in

File: src/com/spikes2212/Cameras/ChangeCamera.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+	public ChangeCamera(CamerasHandler handler, int index) {
+    	
+		super(() -> handler.start(index));
+        // Use requires() here to declare subsystem dependencies
+        // eg. requires(chassis);
+    }
+
+    // Called just before this Command runs the first time
+    protected void initialize() {
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/ChangeCamera.java]: This wouldn't work with all those overrides...

File: src/com/spikes2212/Cameras/ChangeCamera.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
+ */
+public class ChangeCamera extends RunnableCommand {
+
+
+	public ChangeCamera(CamerasHandler handler, int index) {
+
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/ChangeCamera.java]: remove redundant line please

File: src/com/spikes2212/Cameras/StopCameras.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package com.spikes2212.Cameras;
+
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
+ */
+public class StopCameras extends Command {
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/StopCameras.java]: This should also be done with RunnableCommand

File: src/com/spikes2212/genericsubsystems/commands/MonitoredMoveLimitedSubsystem.java [hunk 8]
```diff
@@ -34,7 +34,7 @@ protected void execute() {
 
 	@Override
 	protected boolean isFinished() {
-		return limitedSubsystem.canMove(speedSupplier.get());
+		return (!limitedSubsystem.canMove(speedSupplier.get()))|| isTimedOut();
```
Comment (ThinkRedstone) [inline src/com/spikes2212/genericsubsystems/commands/MonitoredMoveLimitedSubsystem.java]: revert this commit

File: src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java [hunk 9]
```diff
@@ -35,7 +35,7 @@ protected void execute() {
 
 	@Override
 	protected boolean isFinished() {
-		return limitedSubsystem.canMove(speedSupplier.get());
```
Comment (ThinkRedstone) [inline src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java]: revert this commit

File: src/com/spikes2212/utils/RunnableCommand.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    	new Thread(runnable).start();
+    }
+
+    // Called repeatedly when this Command is scheduled to run
+    protected void execute() {
+    }
+
+    // Make this return true when this Command no longer needs to run execute()
+    protected boolean isFinished() {
+        return false;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/utils/RunnableCommand.java]: you want your command to finish at one point or another...

File: src/com/spikes2212/Cameras/CameraController.java [hunk 11]
```diff
@@ -7,7 +7,7 @@
 import edu.wpi.first.wpilibj.vision.USBCamera;
 
 public class CameraController {
-	public static int resolutionWidth = 100, resolutionHeight = 100;
+	public final static int resolutionWidth = 100, resolutionHeight = 100;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CameraController.java]: Please make these values reasonable, check Stronghold project for the defaults we used, those should be fine

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 12]
```diff
@@ -12,15 +12,22 @@
 	private ArrayList<CameraController> cameras;
 	private CameraController currentCamera = null;
 	private Image image;
+	private int currentIndex;
 	public static int fps = 30;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: make this not static, so different camera handlers can have different fps values

File: src/com/spikes2212/Cameras/StopCameras.java [hunk 13]
```diff
@@ -1,39 +1,39 @@
...
-    }
+public class StopCameras extends RunnableCommand {
+
+	public StopCameras(CamerasHandler handler) {
+		super(() -> handler.stop());
+		// Use requires() here to declare subsystem dependencies
+		// eg. requires(chassis);
+	}
+
+	// Called just before this Command runs the first time
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/StopCameras.java]: this wouldn't work with all those overrides

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 14]
```diff
@@ -12,15 +12,22 @@
 	private ArrayList<CameraController> cameras;
 	private CameraController currentCamera = null;
 	private Image image;
+	private int currentIndex;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: You aren't really using this anywhere

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 15]
```diff
@@ -29,7 +36,7 @@ public synchronized void stop() {
 	}
 
 	public synchronized void start(int index) {
-		if (currentCamera != cameras.get(index)) {
+		if (currentIndex != index) {
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: in conjunction with the way addCamera is written, this means you can't start a camera if it's the last one you added, which is problematic

File: src/com/spikes2212/Cameras/ChangeCamera.java [hunk 16]
```diff
@@ -9,33 +9,32 @@
...
-
-    // Called once after isFinished returns true
-    protected void end() {
-    }
-
-    // Called when another command which requires one or more of the same
-    // subsystems is scheduled to run
-    protected void interrupted() {
-    }
+		// Use requires() here to declare subsystem dependencies
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/ChangeCamera.java]: this still doesn't work

File: src/com/spikes2212/genericsubsystems/commands/MonitoredMoveLimitedSubsystem.java [hunk 17]
```diff
@@ -34,7 +34,7 @@ protected void execute() {
 
 	@Override
 	protected boolean isFinished() {
-		return (!limitedSubsystem.canMove(speedSupplier.get()))|| isTimedOut();
+		return limitedSubsystem.canMove(speedSupplier.get());
```
Comment (ThinkRedstone) [inline src/com/spikes2212/genericsubsystems/commands/MonitoredMoveLimitedSubsystem.java]: don't change unrelated files

File: src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java [hunk 18]
```diff
@@ -35,7 +35,7 @@ protected void execute() {
 
 	@Override
 	protected boolean isFinished() {
-		return (!limitedSubsystem.canMove(speedSupplier.get()))|| isTimedOut();
+		return limitedSubsystem.canMove(speedSupplier.get());
```
Comment (ThinkRedstone) [inline src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java]: don't change unrelated files

File: src/com/spikes2212/Cameras/StopCameras.java (hunk 13 above)
Comment (hubiruchi) [inline src/com/spikes2212/Cameras/StopCameras.java]: what overrides?

File: src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java (hunk 18 above)
Comment (hubiruchi) [inline src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java]: fixed

File: src/com/spikes2212/Cameras/CameraController.java (hunk 1 above)
Comment (hubiruchi) [inline src/com/spikes2212/Cameras/CameraController.java]: irrelevant

File: src/com/spikes2212/Cameras/CameraController.java (hunk 2 above)
Comment (hubiruchi) [inline src/com/spikes2212/Cameras/CameraController.java]: irrelevant???

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+import com.ni.vision.NIVision.Image;
+import com.ni.vision.NIVision.ImageType;
+
+import edu.wpi.first.wpilibj.CameraServer;
+
+public class CamerasHandler {
+	private ArrayList<CameraController> cameras;
+	private CameraController currentCamera = null;
+	private Image image;
+	public static int fps = 30;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: Make this non-static

File: src/com/spikes2212/Cameras/CameraController.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
+package com.spikes2212.Cameras;
+
+import java.util.Optional;
+
+import com.ni.vision.NIVision.Image;
+
+import edu.wpi.first.wpilibj.vision.USBCamera;
+
+public class CameraController {
+	public final static int resolutionWidth = 100, resolutionHeight = 100;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CameraController.java]: change to reasonable values, like the ones in the stronghold project

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import com.ni.vision.NIVision;
+import com.ni.vision.NIVision.Image;
+import com.ni.vision.NIVision.ImageType;
+
+import edu.wpi.first.wpilibj.CameraServer;
+
+public class CamerasHandler {
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: should be a subsystem

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+	private Map<String,CameraController> cameras;
+	private CameraController currentCamera = null;
+	private Image image;
+	public int fps = 30;
+
+	public CamerasHandler() {
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: should add a constructor to get camera ports and automatically add them

File: src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java [hunk 23]
```diff
@@ -32,7 +32,6 @@ protected void execute() {
         limitedSubsystem.tryMove(speedSupplier.get());
 
     }
-
```
Comment (ThinkRedstone) [inline src/com/spikes2212/genericsubsystems/commands/MoveLimitedSubsystem.java]: revert this

File: src/com/spikes2212/utils/RunnableCommand.java [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+		new Thread(runnable).start();
+	}
+
+	// Called repeatedly when this Command is scheduled to run
+	protected void execute() {
+	}
+
+	// Make this return true when this Command no longer needs to run execute()
+	protected boolean isFinished() {
+		return true;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/utils/RunnableCommand.java]: return when thread finishes

File: src/com/spikes2212/Cameras/StopCameras.java [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
+ */
+public class StopCameras extends RunnableCommand {
+
+	public StopCameras(CamerasHandler handler) {
+		super(() -> handler.stop());
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/StopCameras.java]: should require the handler

File: src/com/spikes2212/Cameras/ChangeCamera.java [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+import com.spikes2212.utils.RunnableCommand;
+
+
+/**
+ *
+ */
+public class ChangeCamera extends RunnableCommand {
+
+	public ChangeCamera(CamerasHandler handler, String port) {
+
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/ChangeCamera.java]: should require the handler

File: src/com/spikes2212/Cameras/CamerasHandler.java [hunk 27]
```diff
@@ -20,6 +20,8 @@
 	public CamerasHandler() {
 		cameras = new HashMap<>();
 		this.image = NIVision.imaqCreateImage(ImageType.IMAGE_RGB, 0);
+		new Thread(new VisionRunnable(this));
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/CamerasHandler.java]: start the thread...

File: src/com/spikes2212/Cameras/VisionRunnable.java [hunk 28]
```diff
@@ -0,0 +1,21 @@
+package com.spikes2212.Cameras;
+
+import edu.wpi.first.wpilibj.Timer;
+
+public class VisionRunnable implements Runnable{
+	
+	CamerasHandler camerasHandler;
```
Comment (ThinkRedstone) [inline src/com/spikes2212/Cameras/VisionRunnable.java]: make private
